### PR TITLE
#11 SlackTextHandler URL links in text detection implemented

### PR DESCRIPTION
**Pull request by Benkoff**
The task has been completed
connect to #11

### CODE REVIEW

File: links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java [hunk 1]
```diff
@@ -2,17 +2,34 @@
...
+    /**
+     * URL detection is performed by Slack server, returning links surrounded with angle brackets, then
+     * followed by a pipe character (|) and originally typed version of the URL, if it was not taken as-is.
+     * This method returns bookmarkable links: starting with http, https, ftp and ftps prefixes only.
+     */
+    public List<String> getURLsFromText(String text) {
+        List<String> result = new ArrayList<>();
 
-    public static Set<String> getURLsFromText(String text) {
+        log.debug("Searching for URL links...");
```
Comment (ivshapovalov) [inline links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java]: Debug logs at the beginning and end of the method not needed. It's Spring AOP work.

File: links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java [hunk 2]
```diff
@@ -2,17 +2,34 @@
...
+        Pattern p = Pattern.compile("<(ht|f)tp(s?)://.*?([>|])");
+        Matcher matcher = p.matcher(text);
+        while (matcher.find()) {
+            String found = matcher.group().subSequence(1, matcher.group().length() - 1).toString();
+            log.debug("...found: [{}]", found);
```
Comment (ivshapovalov) [inline links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java]: Is this logs required? We can see all result at the end of the method

File: links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java [hunk 3]
```diff
@@ -2,17 +2,34 @@
...
+            result.add(found);
+        }
+        log.debug("URL links found: [{}]", result.size());
```
Comment (ivshapovalov) [inline links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java]: This logs not needed. Spring AOP do this. Maybe replace DEBUG by INFO according to https://github.com/JujaLabs/docs/tree/master/other/workflow

File: links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java [hunk 4]
```diff
@@ -2,17 +2,34 @@
...
  */
 @Slf4j
 public class SlackTextHandler {
+    /**
+     * URL detection is performed by Slack server, returning links surrounded with angle brackets, then
+     * followed by a pipe character (|) and originally typed version of the URL, if it was not taken as-is.
+     * This method returns bookmarkable links: starting with http, https, ftp and ftps prefixes only.
+     */
+    public List<String> getURLsFromText(String text) {
+        List<String> result = new ArrayList<>();
```
Comment (ivshapovalov) [inline links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java:21]: Maybe rename variable as urls?

File: links-slackbot/src/test/java/juja/microservices/links/slackbot/util/SlackTextHandlerTest.java [hunk 5]
```diff
@@ -1,7 +1,46 @@
...
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThat;
+
 /**
- * @author Ivan Shapovalov
+ * @author Ben Novikov
  */
 public class SlackTextHandlerTest {
+    private final SlackTextHandler slackTextHandler = new SlackTextHandler();
```
Comment (ivshapovalov) [inline links-slackbot/src/test/java/juja/microservices/links/slackbot/util/SlackTextHandlerTest.java]: Good practice is to initialize class fields in @Before setUp() method. 
http://projsolution.com/a106-193959-java

File: links-slackbot/src/test/java/juja/microservices/links/slackbot/util/SlackTextHandlerTest.java [hunk 6]
```diff
@@ -1,7 +1,46 @@
...
+
+    @Test
+    public void getURLsFromText_getSet_returnsExpectedSet() {
```
Comment (ivshapovalov) [inline links-slackbot/src/test/java/juja/microservices/links/slackbot/util/SlackTextHandlerTest.java]: Not Set. What does it mean 'getSet'? 
I know MethodName_StateUnderTest_ExpectedBehavior but what is it?

File: links-slackbot/src/test/java/juja/microservices/links/slackbot/util/SlackTextHandlerTest.java [hunk 7]
```diff
@@ -1,7 +1,46 @@
...
+                "Ссылка 9 <ftp://ya.google.mail.ya.au> " +
+                "Ссылка 10 <ftp://name:pass@warehouse.com> " ;
+        List<String> expectedList = new ArrayList<>(Arrays.asList(
+                "http://google.com?parameter=value&also=another",
+                "https://mail.ru",
+                "http://yahoo.us",
+                "ftp://ya.google.mail.ya.au",
+                "ftp://name:pass@warehouse.com"));
+
+        List<String> actualList = slackTextHandler.getURLsFromText(text);
```
Comment (ivshapovalov) [inline links-slackbot/src/test/java/juja/microservices/links/slackbot/util/SlackTextHandlerTest.java]: I think variable name must not contain class name.
Actual and expected is common names. Or maybe 'urls'

File: links-slackbot/src/test/java/juja/microservices/links/slackbot/util/SlackTextHandlerTest.java [hunk 8]
```diff
@@ -1,7 +1,46 @@
...
+
+        assertNotNull(actualList);
```
Comment (ivshapovalov) [inline links-slackbot/src/test/java/juja/microservices/links/slackbot/util/SlackTextHandlerTest.java]: What the purpose of this validation? We initialize new ArrayList at first line of original method. Null is impossilble.

**Review by ivshapovalov — CHANGES_REQUESTED** [review]

**Comment by ivshapovalov** [comment]
New branch for every new task. https://github.com/JujaLabs/docs/tree/master/other/workflow  .Сlause 2.5

File: links-slackbot/src/test/java/juja/microservices/links/slackbot/util/SlackTextHandlerTest.java (hunk 6 above)
Comment (Benkoff) [reply links-slackbot/src/test/java/juja/microservices/links/slackbot/util/SlackTextHandlerTest.java]: oops. old name given to Set<> returning method. thanks

File: links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java (hunk 4 above)
Comment (Benkoff) [reply links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java:21]: Some find that consistently using the conventional name 'result' for the return value of a method can increase legibility. www.javapractices.com/topic/TopicAction.do?Id=63

File: links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java [hunk 9]
```diff
@@ -2,17 +2,32 @@
...
 
-    public static Set<String> getURLsFromText(String text) {
+        Pattern p = Pattern.compile("<(ht|f)tp(s?)://.*?([>|])");
+        Matcher matcher = p.matcher(text);
+        while (matcher.find()) {
+            String found = matcher.group().subSequence(1, matcher.group().length() - 1).toString();
+            log.info("URL link found: [{}]", found);
```
Comment (ivshapovalov) [inline links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java]: Сommon INFO log in the end of method with size of list or list content is bad?

Comment (Benkoff) [reply links-slackbot/src/main/java/juja/microservices/links/slackbot/util/SlackTextHandler.java]: looks like this one provides more details on the objects we discover in progress, so it's definitely more informative. is there any source (besides you mentioned before) to follow data logging rules?

**Review by ivshapovalov — APPROVED** [review]

**Review by adaleorion — APPROVED** [review]
looks good

**Comment by adaleorion** [comment]
@Benkoff 
Please connect pull request to task 
For example, `connect to #11 `

**Pull request merged**